orttraining/tools/scripts/pipeline_model_split.py

```diff
@@ -0,0 +1,376 @@
+import sys
+import onnx
+from onnx import helper
+from onnx import TensorProto
+from onnx import OperatorSetIdProto
+
+# Edge that needs to be cut for the split.
+# If the edge is feeding into more than one nodes, and not all the nodes belong to the same cut,
+# specify those consuming nodes that need to be cut
+class CutEdge:
+    def __init__(self, edgeId, consumingNodes=None):
+        self.edgeId = edgeId
+        self.consumingNodes = consumingNodes
+
+def add_expand_shape(model, name, shape):
+    expand_edge = model.graph.value_info.add()
+    expand_edge.name = name
+    expand_edge.type.CopyFrom(shape)
+
+# Add wait/record/send/recv nodes and split the graph into disconnected subgraphs
+def split_graph(model, edgeIds):
+    ms_domain = 'com.microsoft'
+
+    # Add wait for initial inputs. This needs to be done first before new inputs
+    # are introduced from split
+    initializer_lists = [a.name for a in model.graph.initializer]
+    input_tensors = [
+        value.name for value in model.graph.input if value.name not in initializer_lists]
```

Comment:
>if value.name not in initializer_lists [](start = 50, length = 38)

for ir_version >= 4, if initializer name is in graph_input, it could be overriden by actual input. I think it should not be excluded here.

Reply:
Should we consider the model wrong if a initializer's name is in graph_input? My understanding is that, the graph_input is for data from dataloader, so it won't apply to the initializer's case.

---
In reply to: [400684536](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r400684536) [](ancestors = 400684536)

Reply:
That should work too.

---
In reply to: [401249216](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r401249216) [](ancestors = 401249216,400684536)